Tighten bounds on versions

Current build from hackage fails with

```
[  4 of 234] Compiling Futhark.Pkg.Types ( src/Futhark/Pkg/Types.hs, dist/build/Futhark/Pkg/Types.o, dist/build/Futhark/Pkg/Types.dyn_o )

src/Futhark/Pkg/Types.hs:73:35: error:
    • Couldn't match expected type ‘Maybe T.Text’
                  with actual type ‘[NE.NonEmpty VUnit]’
    • In the pattern: [Str s NE.:| []]
      In the pattern: SemVer 0 0 0 [_] [Str s NE.:| []]
      In an equation for ‘isCommitVersion’:
          isCommitVersion (SemVer 0 0 0 [_] [Str s NE.:| []]) = Just s
   |
73 | isCommitVersion (SemVer 0 0 0 [_] [Str s NE.:| []]) = Just s
   |                                   ^^^^^^^^^^^^^^^^

src/Futhark/Pkg/Types.hs:79:36: error:
    • Couldn't match expected type ‘Maybe T.Text’
                  with actual type ‘[NE.NonEmpty VUnit]’
    • In the fifth argument of ‘SemVer’, namely ‘[Str commit NE.:| []]’
      In the expression:
        SemVer 0 0 0 [Str time NE.:| []] [Str commit NE.:| []]
      In an equation for ‘commitVersion’:
          commitVersion time commit
            = SemVer 0 0 0 [Str time NE.:| []] [Str commit NE.:| []]
   |
79 |   SemVer 0 0 0 [Str time NE.:| []] [Str commit NE.:| []]
   |                                    ^^^^^^^^^^^^^^^^^^^^^

src/Futhark/Pkg/Types.hs:89:66: error:
    • Couldn't match type ‘[f0 VUnit]’ with ‘Maybe T.Text’
      Expected type: ParsecT
                       Void T.Text Data.Functor.Identity.Identity (Maybe T.Text)
        Actual type: ParsecT
                       Void T.Text Data.Functor.Identity.Identity [f0 VUnit]
    • In the second argument of ‘(<*>)’, namely ‘metaData’
      In the expression:
        SemVer <$> majorP <*> minorP <*> patchP <*> preRel <*> metaData
      In an equation for ‘semver'’:
          semver'
            = SemVer <$> majorP <*> minorP <*> patchP <*> preRel <*> metaData
            where
                majorP = digitsP <* char '.'
                minorP = majorP
                patchP = digitsP
                digitsP = read <$> ((T.unpack <$> string "0") <|> some digitChar)
                ....
    • Relevant bindings include
        metaData :: ParsecT
                      Void T.Text Data.Functor.Identity.Identity [f0 VUnit]
          (bound at src/Futhark/Pkg/Types.hs:97:5)
        metaData' :: ParsecT
                       Void T.Text Data.Functor.Identity.Identity (f0 VUnit)
          (bound at src/Futhark/Pkg/Types.hs:98:5)
   |
89 | semver' = SemVer <$> majorP <*> minorP <*> patchP <*> preRel <*> metaData
   |                                                                  ^^^^^^^^
cabal: Failed to build futhark-0.19.4 (which is required by exe:futhark from
```

--- a/cabal.project
+++ b/cabal.project
@@ -1,5 +1,5 @@
 packages: futhark.cabal
-index-state: 2021-03-26T04:41:59Z
+index-state: 2021-04-15T09:02:06Z
 
 package futhark
   ghc-options: -Werror -j

--- a/default.nix
+++ b/default.nix
@@ -20,6 +20,8 @@ let
     packageOverrides = pkgs: rec {
       haskellPackages = pkgs.haskellPackages.override {
         overrides = haskellPackagesNew: haskellPackagesOld: rec {
+          versions =
+            haskellPackagesNew.callPackage ./nix/versions.nix { };
 
           futhark =
             pkgs.haskell.lib.overrideCabal

--- a/futhark.cabal
+++ b/futhark.cabal
@@ -337,7 +337,7 @@ library
     , utf8-string >=1
     , vector >=0.12
     , vector-binary-instances >=0.2.2.0
-    , versions >=4.0.1
+    , versions >=5.0.0
     , zip-archive >=0.3.1.1
     , zlib >=0.6.1.2
   default-language: Haskell2010

--- a/nix/versions.nix
+++ b/nix/versions.nix
@@ -1,0 +1,19 @@
+{ mkDerivation, base, deepseq, hashable, lib, megaparsec, microlens
+, parser-combinators, QuickCheck, tasty, tasty-hunit
+, tasty-quickcheck, text
+}:
+mkDerivation {
+  pname = "versions";
+  version = "5.0.0";
+  sha256 = "f92fed241d1c30ad53d4257a77b7543a07a48667be61d70e370c2bdd7694dde5";
+  libraryHaskellDepends = [
+    base deepseq hashable megaparsec parser-combinators text
+  ];
+  testHaskellDepends = [
+    base megaparsec microlens QuickCheck tasty tasty-hunit
+    tasty-quickcheck text
+  ];
+  homepage = "https://github.com/fosskers/versions";
+  description = "Types and parsers for software version numbers";
+  license = lib.licenses.bsd3;
+}

--- a/src/Futhark/Pkg/Types.hs
+++ b/src/Futhark/Pkg/Types.hs
@@ -70,13 +70,13 @@ pkgPathFilePath = joinPath . Posix.splitPath . T.unpack
 -- @hash@ (typically the Git commit ID).  This function detects such
 -- versions.
 isCommitVersion :: SemVer -> Maybe T.Text
-isCommitVersion (SemVer 0 0 0 [_] [Str s NE.:| []]) = Just s
+isCommitVersion (SemVer 0 0 0 [_] (Just s)) = Just s
 isCommitVersion _ = Nothing
 
 -- | @commitVersion timestamp commit@ constructs a commit version.
 commitVersion :: T.Text -> T.Text -> SemVer
 commitVersion time commit =
-  SemVer 0 0 0 [Str time NE.:| []] [Str commit NE.:| []]
+  SemVer 0 0 0 [Str time NE.:| []] (Just commit)
 
 -- | Unfortunately, Data.Versions has a buggy semver parser that
 -- collapses consecutive zeroes in the metadata field.  So, we define
@@ -94,8 +94,8 @@ semver' = SemVer <$> majorP <*> minorP <*> patchP <*> preRel <*> metaData
     digitsP = read <$> ((T.unpack <$> string "0") <|> some digitChar)
     preRel = maybe [] pure <$> optional preRel'
     preRel' = char '-' *> (pure . Str . T.pack <$> some digitChar)
-    metaData = maybe [] pure <$> optional metaData'
-    metaData' = char '+' *> (pure . Str . T.pack <$> some alphaNumChar)
+    metaData = optional metaData'
+    metaData' = char '+' *> (T.pack <$> some alphaNumChar)
 
 -- | The dependencies of a (revision of a) package is a mapping from
 -- package paths to minimum versions (and an optional hash pinning).

--- a/stack.yaml
+++ b/stack.yaml
@@ -15,3 +15,6 @@ nix:
 # This makes it do so.
 ghc-options:
   $locals: -j -optP-Wno-nonportable-include-path
+
+extra-deps:
+- versions-5.0.0

